### ISSUE TITLE
pkg/errors is no longer maintained and archived

### ISSUE THREAD

**Issue by sbueringer** [kubernetes-sigs/cluster-api#6688]
To be honest, I'm not sure what we want to do or if we want to do something, but just found out that pkg/errors is archived.

See: https://github.com/pkg/errors/issues/245

/kind bug
/area dependency
[One or more /area label. See https://github.com/kubernetes-sigs/cluster-api/labels?q=area for the list of labels]

**Comment by DiptoChakrabarty** [comment]
There are two options which I think can be done

- keep things as they are
- replace with native golang error wrapping which can be handled by the fmt and errors package

there might be more but these come to my mind

there are few issues regarding this in other repos  https://github.com/kubernetes/release/issues/2549

**Comment by oscr** [comment]
@sbueringer What are your thoughts on this? Is this something we can start working on, but wait with prs until 1.2 is released?

**Comment by sbueringer** [comment]
I think we need more discussion before we should do anything.

I would especially like to here the opinions of other maintainers on this as it's a pretty fundamental change

(@enxebre @CecileRobertMichon @fabriziopandini @vincepri)

**Comment by fabriziopandini** [comment]
I would align to whatever k/k does
@neolit123 any info WRT to ^^

**Comment by neolit123** [comment]
the Wrap(f) gives formatting consistency and avoids leakage of `text: ...` vs `text : ...` formatting diffs. stdlib has no answer.

k/k is not taking action for kubeadm the time being, although there was a ticket somewhere.
unless k/k/kubeadm is forced to switch we can continue vendoring it.

**Comment by fabriziopandini** [comment]
/triage accepted

Let's keep this around to monitor the situation, but I will aling to k/k and not take action for the time being

**Comment by fabriziopandini** [comment]
(doing some cleanup on old issues without updates)
/close
unfortunately, no one is picking up the task. the thread will remain available for future reference

**Comment by sbueringer** [comment]
/reopen

Given that I see continued comments from maintainers if we should get rid of pkg/errors

**Comment by sbueringer** [comment]
The problem with just directly replacing it through Go SDK '%w' is that we lose information. Most visible in logs.

Example:
```go
	err1 := fmt.Errorf("err1")
	err2 := fmt.Errorf("err2: %w", err1)
	log.Error(err2, "Go SDK error wrapping")

	err1 = pkgerrors.New("err1")
	err2 = pkgerrors.Wrapf(err1, "err2")
	log.Error(err2, "pkgerrors wrapping")
```

Resulting logs
```
{"ts":1706257281192.585,"caller":"cluster-api/main.go:252","msg":"Go SDK error wrapping","err":"err2: err1"}
{"ts":1706257281192.6028,"caller":"cluster-api/main.go:256","msg":"pkgerrors wrapping","err":"err2: err1","errVerbose":"err1\nmain.main\n\t/Users/buringerst/code/src/sigs.k8s.io/cluster-api/main.go:254\nruntime.main\n\t/Users/buringerst/go_sdks/go1.21.5/src/runtime/proc.go:267\nruntime.goexit\n\t/Users/buringerst/go_sdks/go1.21.5/src/runtime/asm_arm64.s:1197\nerr2\nmain.main\n\t/Users/buringerst/code/src/sigs.k8s.io/cluster-api/main.go:255\nruntime.main\n\t/Users/buringerst/go_sdks/go1.21.5/src/runtime/proc.go:267\nruntime.goexit\n\t/Users/buringerst/go_sdks/go1.21.5/src/runtime/asm_arm64.s:1197"}
```
Note: Full error stack traces are only implemented with JSON format not with text. This is because JSON logging in component-base is implemented with zap and zap automatically adds errVerbose if err implements fmt.Formatter (xref: https://github.com/uber-go/zap/blob/master/zapcore/error.go#L70-L77)

As can be seen, pkg/errors preserves the stack trace of where the error occurs which can vastly simplify debugging.

I think we have the following options:
* lose information and move to Go SDK wrapping
* implement our own util package which just preserves the stacktrace and is otherwise a 1:1 replacement for pkg/errors (not sure how much code we're talking about)
* keep pkg/errors (not great, given that it's unmaintained)

**Comment by neolit123** [comment]
> keep pkg/errors (not great, given that it's unmaintained)

if it's no longer maintainer, but it has no critical bugs, then there is no problem with using it?

> implement our own util package which just preserves the stacktrace and is otherwise a 1:1 replacement for pkg/errors (not sure how much code we're talking about)

maybe we can try to add it in component-base?
https://github.com/kubernetes/component-base
needs sig-arch approval, but makes more sense.

or kubernetes/kubeadm/errors?
@fabriziopandini 

> lose information and move to Go SDK wrapping

and lose formatting, since sometimes `err2 := fmt.Errorf("err2:   %w", err1)` can be missed during review.

**Comment by sbueringer** [comment]
> if it's no longer maintainer, but it has no critical bugs, then there is no problem with using it?

Good point. I think there is no reason why we would have to migrate away now.


I mostly reopened this because it was brought up  a few times on PRs and this issue seems like a better place for discussion.

**Comment by fabriziopandini** [comment]
I don't think that as a CAPI community w should/we have enough resources to maintain an error package
I would simply align and what k/k does.

**Comment by neolit123** [comment]
> I would simply align at what k/k does.

in k/k there was a tracking issue to drop all usage, but for cmd/kubeadm we requested an exception.

**Comment by neolit123** [comment]
FYI, that wasn't urgent but kubeadm switched to a local implementation of wrapped errors with stack traces.
https://github.com/kubernetes/kubernetes/pull/132080/commits/f522d7cb858020116a572d5610d6b3f72c32a50f#diff-a355f64adf9229aaa64b22e80b446081cc728839cb426d799145cfef4ba8fa9eR1

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/util/errors/errors.go#L41

**Comment by sbueringer** [comment]
While looking into this migration we should also figure out if we want to continue to wrap errors roughly everywhere. Currently this leads to very verbose logs if the error stack is logged (e.g. when --format=json is used), but maybe that is intended?

**Comment by neolit123** [comment]
> While looking into this migration we should also figure out if we want to continue to wrap errors roughly everywhere. Currently this leads to very verbose logs if the error stack is logged (e.g. when --format=json is used), but maybe that is intended?

WithMessage is suitable to reduce the wrapping.
